internals/overlord/logstate/loki/loki.go

```diff
@@ -0,0 +1,203 @@
+	}
+
+	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, c.remoteURL, bytes.NewReader(jsonReq))
+	if err != nil {
+		return fmt.Errorf("creating HTTP request: %v", err)
+	}
+	httpReq.Header.Set("Content-Type", "application/json; charset=utf-8")
+	httpReq.Header.Set("User-Agent", fmt.Sprintf("pebble/%s", cmd.Version))
+
+	resp, err := c.httpClient.Do(httpReq)
+	if err != nil {
+		return err
+	}
+
+	err, emptyBuffer := c.handleServerResponse(resp)
+	if emptyBuffer {
+		if err != nil {
+			logger.Noticef("Target %q: request failed, dropping %d logs",
+				c.targetName, c.numEntries)
+		}
+		c.emptyBuffer()
+	}
+	return err
+}
+
+func (c *Client) emptyBuffer() {
+	for svc := range c.entries {
+		c.entries[svc] = c.entries[svc][:0]
+	}
+	c.numEntries = 0
+}
+
+func (c *Client) buildRequest() request {
+	// Sort keys to guarantee deterministic output
+	services := make([]string, 0, len(c.entries))
+	for svc, entries := range c.entries {
+		if len(entries) == 0 {
+			continue
+		}
+		services = append(services, svc)
+	}
+	sort.Strings(services)
+
+	var req request
+	for _, service := range services {
+		entries := c.entries[service]
+		stream := stream{
+			Labels: map[string]string{
+				"pebble_service": service,
+			},
+			Entries: entries,
+		}
+		req.Streams = append(req.Streams, stream)
+	}
+	return req
+}
+
+type request struct {
+	Streams []stream `json:"streams"`
+}
+
+type stream struct {
+	Labels  map[string]string `json:"stream"`
+	Entries []lokiEntry       `json:"values"`
+}
+
+type lokiEntry [2]string
+
+// handleServerResponse determines what to do based on the response from the
+// Loki server. 4xx and 5xx responses indicate errors, so in this case, we will
+// bubble up and error to the caller. The returned boolean indicates whether
+// the buffer should be emptied.
+func (c *Client) handleServerResponse(resp *http.Response) (err error, emptyBuffer bool) {
+	defer func() {
+		// Drain request body to allow connection reuse
+		// see https://pkg.go.dev/net/http#Response.Body
+		_, _ = io.Copy(io.Discard, io.LimitReader(resp.Body, 1024*1024))
+		_ = resp.Body.Close()
+	}()
+
+	code := resp.StatusCode
+	switch {
+	case code == http.StatusTooManyRequests:
+		// For 429, don't drop logs - just retry later
+		return errFromResponse(resp), false
+
+	case code >= 500:
+		// 5xx indicates a problem with the server, so don't drop logs
+		return errFromResponse(resp), false
+
+	case code >= 400:
+		// 4xx indicates a client problem, so drop the logs
+		return errFromResponse(resp), true
+	}
+
+	return nil, true
+}
+
+// errFromResponse generates an error from a failed *http.Response.
+// NB: this function reads the response body.
+func errFromResponse(resp *http.Response) error {
+	// Request to Loki failed
+	errStr := fmt.Sprintf("server returned HTTP %d\n", resp.StatusCode)
+
+	// Read response body to get more context
+	body := make([]byte, 0, 1024)
+	_, err := resp.Body.Read(body)
```

Comment:
I don't think this does what you intend -- it will always read 0 bytes. The `body` slice has length 0 (cap 1024) and [`Read` reads](https://pkg.go.dev/io#Reader) len bytes, not cap bytes. It also can't change the slice length because the slice is passed to `Read` by value. So you're almost always better off calling `io.ReadAll` or similar. I'd suggest:

```go
body, err := io.ReadAll(LimitReader(resp.Body, 1024))
```